lib/spack/spack/binary_distribution.py

```diff
@@ -631,3 +708,114 @@ def get_keys(install=False, trust=False, force=False):
+        raise spack.error.SpackError(msg)
+    except Exception as expn:
+        msg = 'Error getting file contents at {0} due to {1}'.format(
+            file_uri, expn.message)
+        raise spack.error.SpackError(msg)
+
+    file_contents = url.read()
+
+    if not file_contents:
+        msg = 'Unable to read file contents at {0}'.format(file_uri)
+        raise spack.error.SpackError(msg)
+
+    return file_contents
+
+
+def needs_rebuild(spec, mirror_url, buildcache_index):
+    pkg_name = spec.name
+    pkg_version = spec.version
+
+    tty.msg('Checking {0}-{1}'.format(pkg_name, pkg_version))
+
+    spec.concretize()
+    pkg_hash = spec.dag_hash()
+    pkg_full_hash = spec.full_hash()
+
+    rebuild_spec = {
+        'short_spec': spec.short_spec,
+        'hash': pkg_hash
+    }
+
+    if buildcache_index:
+        # just look in the index we already fetched
+        if pkg_hash in buildcache_index:
+            # At least remote binary mirror knows about it, so if the
+            # full_hash doesn't match (or remote end doesn't know about
+            # the full_hash), then we trigger a rebuild.
+            remote_pkg_info = buildcache_index[pkg_hash]
+            if ('full_hash' not in remote_pkg_info or
+                remote_pkg_info['full_hash'] != pkg_full_hash):
+                    return rebuild_spec
+        else:
+            # remote binary mirror doesn't know about this package, we
+            # should probably just rebuild it
+            return rebuild_spec
+    else:
+        # retrieve the .spec.yaml and look there instead
+        build_cache_dir = build_cache_directory(mirror_url)
+        spec_yaml_file_name = tarball_name(spec, '.spec.yaml')
+        file_path = os.path.join(build_cache_dir, spec_yaml_file_name)
+        try:
+            yaml_contents = read_from_url(file_path)
+        except spack.error.SpackError:
+            # let any kind of failure reading the .spec.yaml indicate rebuild
+            return rebuild_spec
+        spec_yaml = syaml.load(yaml_contents)
+
+        if ('full_hash' not in spec_yaml or
+            spec_yaml['full_hash'] != pkg_full_hash):
+                return rebuild_spec
+
+    return None
+
+
+def get_remote_index(mirror_url):
+    build_cache_dir = build_cache_directory(mirror_url)
+
+    # First fetch the index.json
+    index_path = os.path.join(build_cache_dir, 'index.json')
+    index_contents = read_from_url(index_path)
+
+    return json.loads(index_contents)
+
+
+def check(mirrors, specs, no_index=False, output_file=None):
+    rebuilds = {}
+    for mirror in mirrors.keys():
+        mirror_url = mirrors[mirror]
+        tty.msg('Checking for built specs at %s' % mirror_url)
+
+        rebuild_list = []
+        remote_pkg_index = None
+        if not no_index:
+            remote_pkg_index = get_remote_index(mirror_url)
+
+        for spec in specs:
+            rebuild_spec = needs_rebuild(spec, mirror_url, remote_pkg_index)
```

Comment:
I wonder if this could be made faster with logic like `spack.util.web._spider` or just its own `multithreading.Pool` -- right now it does a bunch of potentially slow web requests sequentially.